server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -2446,18 +2478,83 @@ private void deleteSnapshotsFromRepository(
+                                removeSnapshotDeletionFromClusterState(deleteEntry, e, repositoryData);
+                            }
+                        }
+                    } finally {
+                        latch.countDown();
+                    }
+                });
+                try {
+                    latch.await();
+                    RemoteSegmentStoreDirectoryFactory remoteSegmentStoreDirectoryFactory = new RemoteSegmentStoreDirectoryFactory(
+                        () -> repositoriesService,
+                        threadPool
+                    );
+                    if (snapshotsWithLockFiles.size() > 0) {
+                        repository.deleteSnapshotsAndReleaseLockFiles(
+                            snapshotsWithLockFiles,
+                            repositoryData.getGenId(),
+                            minCompatibleVersion(minNodeVersion, repositoryData, snapshotsWithLockFiles),
+                            remoteStoreLockManagerFactory,
+                            remoteSegmentStoreDirectoryFactory,
+                            false,
+                            ActionListener.wrap(updatedRepoData -> {
+                                logger.info("snapshots {} deleted", snapshotsWithLockFiles);
+                                removeSnapshotDeletionFromClusterState(deleteEntry, null, updatedRepoData);
+                            }, ex -> removeSnapshotDeletionFromClusterState(deleteEntry, ex, repositoryData))
+                        );
+                    }
+                    if (snapshotsWithPinnedTimestamp.size() > 0) {
+
+                        final StepListener<RepositoryData> pinnedTimestampListener = new StepListener<>();
+                        pinnedTimestampListener.whenComplete(
+                            updatedRepoData -> { removeSnapshotDeletionFromClusterState(deleteEntry, null, updatedRepoData); },
+                            ex -> removeSnapshotDeletionFromClusterState(deleteEntry, ex, repositoryData)
+                        );
+
+                        repository.deleteSnapshotsWithPinnedTimestamp(
+                            snapshotsWithPinnedTimestamp.keySet(),
+                            repositoryData.getGenId(),
+                            minCompatibleVersion(minNodeVersion, repositoryData, snapshotsWithPinnedTimestamp.keySet()),
+                            remoteSegmentStoreDirectoryFactory,
+                            ActionListener.wrap(updatedRepoData -> {
+                                logger.info("snapshots {} deleted", snapshotsWithPinnedTimestamp);
+                                removeSnapshotsPinnedTimestamp(
+                                    snapshotsWithPinnedTimestamp,
+                                    repository,
+                                    updatedRepoData,
+                                    pinnedTimestampListener
+                                );
```

Comment:
shouldn't this be done before `deleteSnapshotsWithPinnedTimestamp` is done ? As otherwise the deleted indices would still the pinned timestamp and do nothing to clean the remote store directory.

Reply:
modified the code to unpin timestamp after repository data update